### ISSUE TITLE
Cacti allows the removal of Data Query Graph Template associations when they are in use

### ISSUE THREAD

**Issue by DavidLiedke** [Cacti/cacti#521]
I use Cacti 1.1.2

Console --> Management --> Graphs --> List of Graphs. 

![unbenannt](https://cloud.githubusercontent.com/assets/25581728/24694942/23b5b3dc-19e4-11e7-87d0-4b15cf7395f7.png)

I get wrong Templates for the graphs. 

If i select one of these graphes. I get no value for the graph template:

![unbenannt2](https://cloud.githubusercontent.com/assets/25581728/24695066/9780abc8-19e4-11e7-8573-52caac8074bd.png)

**Comment by cigamit** [comment]
Does it appear that way when you unlock the graph?  If it's still reporting a non-templated value, please check the clog for errrors if you dont find any, then do the following:

1) Note the local graph id.
2) SELECT * FROM graph_local WHERE id = <yourgraphid>
3) SELECT * FROM graph_templates WHERE id = <thegraph_template_id_from_above>
4) SELECT * FROM snmp_query_graph WHERE snmp_query_graph_id = <snmp_query_graph_id_from_above>

Let us know what is returned in each case, or the SQL Error otherwise.

**Comment by cigamit** [comment]
David, any progress?

**Comment by DavidLiedke** [comment]
Sorry for the late answer.

I didn´t find anything in the log.

1. 14
2. 
![firefox_2017-04-07_08-45-52](https://cloud.githubusercontent.com/assets/25581728/24788595/a1f3e3aa-1b6e-11e7-8ebb-46885bb50b5d.png)
3.
![firefox_2017-04-07_08-47-00](https://cloud.githubusercontent.com/assets/25581728/24788617/c4b660b6-1b6e-11e7-8dd6-a655abe5ca9e.png)
4. 
![firefox_2017-04-07_08-53-10](https://cloud.githubusercontent.com/assets/25581728/24788807/c5e489a8-1b6f-11e7-8e1e-12f112b55441.png)

**Comment by cigamit** [comment]
Sorry, the last one was more like: 

SELECT * FROM snmp_query_graph WHERE id = 38;

**Comment by DavidLiedke** [comment]
It get a empty result. 

There is no entry in this table for id 38.

**Comment by cigamit** [comment]
This explain's it.  When you edit a Data Query you have the ability to Map the Query Elements to a Graph Template.  In your case, that Graph Template was removed from the mapping and just checking, we are still allowing this, which I'm going to say is the real bug.

Basically, when a Data Query Graph Template is in use (Has Graphs), it should not be allowed to be removed, only edited.

So, in your case, you will need to add back the mapping, note the "Mapping ID", and then, in the table: "graph_local", do the following changes.

UPDATE graph_local SET snmp_query_graph_id="newid" WHERE snmp_query_graph_id=38;

The old Cacti 0.x.x had a number of these bad behaviors and we have tried to cork most of them up.  It seems we forgot one hole.

Does this add up for you?

**Comment by DavidLiedke** [comment]
I had the Graph Template "IOS Interface Traffic" with use the Data Queries "SNMP - Interface Statistics". I had a lot of graphes which use this Graph template. Without any problems. 

All graphes in the "graph_local" table with the "graph_template_id=38" are using the "snmp_query_id=1" ("SNMP - Interface Statistics"). I think this is good.

We didn´t change anything in the past at cacti. The only change was to update it. 

If i open the graph in the "Management-Graphs" section i get the following messages in the cacti log:
```
2017-04-08 17:53:01 - DBCALL ERROR: SQL Assoc Failed!, Error: You have an error in your SQL syntax; check the manual that corresponds to your MariaDB server version for the right syntax to use near ')) OR graph_template_id IN () ORDER BY name' at line 1
2017-04-08 17:53:01 - DBCALL ERROR: SQL Assoc Failed!, Error:1064, SQL:"SELECT CONCAT_WS('', graph_template_id, '_', id, '') AS id, name FROM snmp_query_graph WHERE (snmp_query_id = 1 AND id IN ()) OR graph_template_id IN () ORDER BY name"
```

**Comment by cigamit** [comment]
Hmm.  Maybe we missed something.  I have marked this issue resolved, but your case needs more attention relative to the SQL error.  We will need you to be a developer here.  It almost appears that there are more issues associated with your setup.  If you could edit graphs.php, and look at the code block between lines 407 and 433.  What it tries to do if find common Graph Templates to the one's in question.  In your case, due to either empti $sqgi, $query_fields, $common_graph_ids, you are getting the error.  Can you please determine which it is.

**Comment by DavidLiedke** [comment]
Sorry i am no developer.

I looked into a old database backup of our installation with version 0.8.8h. There was no matching between the Data Query and the Graph Template. If you want i can give you an dump of our database with cacti version 1.1.2 and the backup with version 0.8.8h. 

graphs.php lines between 406 and 434:

[graphs_406_434.pdf](https://github.com/Cacti/cacti/files/909069/graphs_406_434.pdf)

**Comment by DavidLiedke** [comment]
I add the match between the Graph template and the SNMP Data Query. After that i have no "DBCALL Error" anymore. 

After updating the graphs to the new snmp_query_graph_id
[Your SQL statement](https://github.com/Cacti/cacti/issues/521#issuecomment-292711596)
i see in the "Management - Graphs" Section the correct Graph template. 
![chrome_2017-04-10_20-26-57](https://cloud.githubusercontent.com/assets/25581728/24876478/2c85f4d0-1e2c-11e7-8d43-149231269c28.png)

But if i open one graph the wrong graph template is listed. 
![chrome_2017-04-10_20-27-28](https://cloud.githubusercontent.com/assets/25581728/24876487/313eedb0-1e2c-11e7-8441-ad067ef4fca8.png)

**Comment by cigamit** [comment]
Yea, you screwed up.  It might be good to get the dump to a location where it can be had.  This is not a support forum, but I am interested from a tool perspective as to what has caused this to begin with.  You can sent a PM to me on the Cacti forums with instructions on how to obtain it.

**Comment by DavidLiedke** [comment]
I get the message: "You are not authorised to read private messages."

**Comment by cigamit** [comment]
What is your forum account name?

**Comment by DavidLiedke** [comment]
Sorry. Its the same like Github --> DavidLiedke

**Comment by cigamit** [comment]
You should be able to now.  Thanks.

**Comment by DavidLiedke** [comment]
Sorry i always get the same error message.

**Comment by cigamit** [comment]
I was unable to make the permissions change stick either.  Why don't you drop it somewhere where it can be picked up/downloaded.

**Comment by DavidLiedke** [comment]
0.8.8h
https://www.hidrive.strato.com/lnk/UeCMt0BD
1.1.2
https://www.hidrive.strato.com/lnk/GrCstcDS

**Comment by cigamit** [comment]
Got them.  Will take a look in the next few days.

**Comment by cigamit** [comment]
David, There was a 'tuning' step done by reboot a week or so ago that likely fixed this issue.  Should be released in 1.1.5.  Let me know if you are still having the issue.

**Comment by cigamit** [comment]
pretty sure the issue in the title is resolved
